File: LeetCode/dp/HouseRobberIII.java
package dp;

public class HouseRobberIII {
    public int rob(TreeNode root) {
        int[] result = robSubtree(root);
        return Math.max(result[0],result[1]);
    }

    private int[] robSubtree(TreeNode node) {
       if (node == null) {
           return new int[]{0, 0};
       }

       int[] left = robSubtree(node.left);
       int[] right = robSubtree(node.right);

       int rob = node.val + left[1] + right[1];
       int notRob = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);

       return new int[]{rob, notRob};
    }

    public static void main(String[] args) {
        HouseRobberIII solution = new HouseRobberIII();

        // Test case
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.right = new TreeNode(3);
        root.right.right = new TreeNode(1);

        System.out.println(solution.rob(root)); // Output: 7
    }

    // Definition for a binary tree node.
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }
}

File: LeetCode_Python/dp/houserobber/HouseRobberIII.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class HouseRobberIII:
    def rob(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node:
                return (0, 0)

            left = dfs(node.left)
            right = dfs(node.right)

            rob = node.val + left[1] + right[1]
            not_rob = max(left) + max(right)

            return (rob, not_rob)

        return max(dfs(root))


# Time Complexity: O(n), where n is the number of nodes in the tree
# Space Complexity: O(h), where h is the height of the tree (due to recursion stack)

if __name__ == "__main__":
    solution = HouseRobberIII()

    root = TreeNode(3)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.right = TreeNode(3)
    root.right.right = TreeNode(1)
    print("Expected Output: 7")
    print("Actual Output:", solution.rob(root))

    root2 = TreeNode(3)
    root2.left = TreeNode(4)
    root2.right = TreeNode(5)
    root2.left.left = TreeNode(1)
    root2.left.right = TreeNode(3)
    root2.right.right = TreeNode(1)
    print("Expected Output: 9")
    print("Actual Output:", solution.rob(root2))
